### PR TITLE
added is synced check to hare, block creation, atx creation and block…

### PR DESCRIPTION
**Pull request by antonlerner**
… gossip

### CODE REVIEW

File: sync/block_listener.go [hunk 1]
```diff
@@ -65,6 +65,10 @@ func (bl *BlockListener) ListenToGossipBlocks() {
 			bl.Log.Info("listening  stopped")
 			return
 		case data := <-bl.receivedGossipBlocks:
+			if !bl.IsLatest() {
```
Comment (almogdepaz) [inline sync/block_listener.go]: we check this later in the method, we also decided to keep these blocks in the database (no propagating)
so we don't have to fetch them again

File: sync/syncer.go [hunk 2]
```diff
@@ -83,6 +83,11 @@ func (s *Syncer) IsSynced() bool {
 	return s.ValidatedLayer()+1 == s.maxSyncLayer()
 }
 
+func (s *Syncer) IsLatest() bool {
```
Comment (almogdepaz) [inline sync/syncer.go]: better to change is synced logic since  it is used in some places
this is basically the same method except the >=

File: activation/activation.go [hunk 3]
```diff
@@ -126,6 +128,10 @@ func (b *Builder) loop() {
 		case <-b.stop:
 			return
 		case layer := <-b.timer:
+			if !b.isSynced() {
+				b.log.Error("cannot create atx : not synced")
```
Comment (zalmen) [inline activation/activation.go]: It's not an error, change to info (I want to reduce the number of errors/warnings so that eventually we can check at the end of every test that the number of errors/warnings is 0)

File: sync/syncer.go (hunk 2 above)
Comment (antonlerner) [reply sync/syncer.go]: notice that this method uses the LatestLayer and not VerifiedLayer, I'm not sure this is the functionality expected from `IsSynced` currently

Comment (almogdepaz) [reply sync/syncer.go]: Np, we need one way to do it you can change it if it works

Comment (almogdepaz) [reply sync/syncer.go]: Np, we need one way to do it you can change it if it works

Comment (zalmen) [reply sync/syncer.go]: What are the diffrences between `ValidatedLayer` and `LatestLayer`?

File: sync/block_listener.go (hunk 1 above)
Comment (y0sher) [reply sync/block_listener.go]: what if they're bad blocks? our db could be flooded

**Comment by y0sher** [comment]
maybe add a few negative tests where `isSynced` returns false..

File: hare/hare.go [hunk 4]
```diff
@@ -148,6 +152,10 @@ func (h *Hare) onTick(id types.LayerID) {
 		h.lastLayer = id
 	}
 	h.layerLock.Unlock()
+	if !h.isSynced() {
```
Comment (zalmen) [inline hare/hare.go]: REmove this, @gavraz is working on a PR that better handle this case (there are more issues regarding early messages)

Comment (antonlerner) [reply hare/hare.go]: should we remove this or merge this code and then later revise according to @gavraz PR?

File: sync/syncer.go (hunk 2 above)
Comment (antonlerner) [reply sync/syncer.go]: `Validated` layer is the last layer validated by the tortoise, `Latest` is the last layer seen on gossip/ sync

File: sync/block_listener.go (hunk 1 above)
Comment (zalmen) [inline sync/block_listener.go]: Why not use `IsSynced`?

Comment (almogdepaz) [reply sync/block_listener.go]: its an optimisation traffic 
there are ways of handling the problem you suggested like limiting the size

File: sync/syncer.go (hunk 2 above)
Comment (almogdepaz) [reply sync/syncer.go]: yuval made me look at this again and i think that if you want to shut down the listener until synced just use isSynced the condition there is even harder.
you can use it right after receivedGossipBlocks

File: hare/hare.go (hunk 4 above)
Comment (zalmen) [reply hare/hare.go]: hmm I wanted to ease on Gavrad's merge, but either way is fine

File: sync/syncer.go (hunk 2 above)
Comment (antonlerner) [reply sync/syncer.go]: I've changed the synched parameter after talking to Yuval, I think ths=is fits here best, @zalmen  can you please clarify?

File: hare/hare.go (hunk 4 above)
Comment (antonlerner) [reply hare/hare.go]: I've removed this call and the api should remain the same

File: sync/block_listener.go (hunk 1 above)
Comment (antonlerner) [reply sync/block_listener.go]: IsSynched = IsLatest

File: sync/syncer.go (hunk 2 above)
Comment (antonlerner) [reply sync/syncer.go]: IsSynced == IsLatest, see my fixes

File: sync/syncer.go [hunk 5]
```diff
@@ -130,6 +131,7 @@ func NewSync(srv service.Service, layers *mesh.Mesh, txpool MemPool, atxpool Mem
 		txpool:         txpool,
 		atxpool:        atxpool,
 		startLock:      0,
+		currentLayer:   currentLayer,
```
Comment (almogdepaz) [inline sync/syncer.go:134]: why is this needed ? we are getting the current layer from the clock we subscribed to 
we shouldn't have to add this

Comment (almogdepaz) [reply sync/syncer.go:134]: if the issue is the time until the next tick we should receive the clock to the contractor and then set the logic and the subscription in here

**Review by almogdepaz — APPROVED** [review]

File: sync/syncer.go (hunk 5 above)
Comment (antonlerner) [reply sync/syncer.go:134]: I'll add a technical debt, it seems that tests are failing because some assumptions regarding the channels the syncers are listening to are not correct, we need to revise the tests before making this change

Comment (antonlerner) [reply sync/syncer.go:134]: #1043
